storage/s3/src/main/java/io/aiven/kafka/tieredstorage/storage/s3/S3MultiPartOutputStream.java

```diff
@@ -122,31 +142,31 @@ public void close() throws IOException {
         closed = true;
     }
 
-    private void flushBuffer(final int offset,
-                             final int actualPartSize) throws IOException {
+    private void flushBuffer(final ByteBuffer partBuffer, final int actualPartSize) throws IOException {
         try {
-            final ByteArrayInputStream in = new ByteArrayInputStream(partBuffer.array(), offset, actualPartSize);
-            uploadPart(in, actualPartSize);
-            partBuffer.clear();
+            final byte[] array = new byte[actualPartSize];
+            partBuffer.get(array, 0, actualPartSize);
+
+            final var uploadPartRequest = new UploadPartRequest()
+                .withBucketName(bucketName)
+                .withKey(key)
+                .withUploadId(uploadId)
+                .withPartSize(actualPartSize)
+                .withPartNumber(partNumber.incrementAndGet())
+                .withInputStream(new ByteArrayInputStream(array));
+
+            // Run request async
+            partUploads = partUploads.thenCombine(
+                CompletableFuture.supplyAsync(() -> client.uploadPart(uploadPartRequest)),
+                (partETags, result) -> {
+                    partETags.put(result.getPartETag().getPartNumber(), result.getPartETag().getETag());
```

Comment:
I was thinking a little further and because of this PartETag reconstructing can we maybe use ConcurrentLinkedQueue here? Then when passing the results further to AWS client we can just use ArrayList constructor that is able to accept this queue as an argument

Reply:
Good idea, I like it. Applying the suggestion